### PR TITLE
Add user-impersonation support to the Java JobClient

### PR DESCRIPTION
**Pull request by DaoWen**
## Changes proposed in this PR

- Add impersonation support to the Java JobClient.
- Add `interface JobClientInterface` and `abstract class JobClientProxy` to support the new impersonation proxy interface to the JobClient.

## Why are we making these changes?

Users of the JobClient want to take advantage of our user impersonation support. This is done by calling `jobClient.impersonating("target-user")` on an existing JobClient instance, and then making normal calls on the returned proxy class. For example:

```java
jobClient.impersonating("target-user").submit(jobs);
```

### CODE REVIEW

File: jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/Constraints.java [hunk 1]
```diff
@@ -55,7 +56,8 @@ public static Constraint buildEqualsConstraint(String attribute, String value) {
      *                   See {@link Constraint}.
      * @return the parsed constraint.
      */
-    public static Constraint parseFrom(JSONArray constraint) {
+    public static Constraint parseFrom(JSONArray constraint)
+        throws JSONException {
```
Comment (DaoWen) [inline jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/Constraints.java]: @WenboZhao asked for me to add this

File: jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/OneToOneConstraint.java [hunk 2]
```diff
@@ -58,7 +59,8 @@ public int hashCode() {
     }
 
     @Override
-    public JSONArray toJson() {
+    public JSONArray toJson()
+        throws JSONException {
```
Comment (DaoWen) [inline jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/OneToOneConstraint.java]: @WenboZhao asked for me to add this

File: jobclient/src/main/java/com/twosigma/cook/jobclient/Job.java [hunk 3]
```diff
@@ -556,6 +557,18 @@ public Builder setProgressRegexString(String progressRegexString) {
...
+         * Set the user of the job expected to build.
+         *
+         * @param user {@link String} specifies the user for a job.
+         * @return this builder.
+         */
+        public Builder setUser(String user) {
+            _user = user;
+            return this;
+        }
+
```
Comment (pschorf) [inline jobclient/src/main/java/com/twosigma/cook/jobclient/Job.java]: nitpick: extra newline

File: jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java [hunk 4]
```diff
@@ -1113,4 +1103,76 @@ public String toString() {
                 + ", _activeUUIDToJob=" + _activeUUIDToJob + ", _batchSubmissionLimit=" + _batchRequestSize
                 + ", _statusUpdateInterval=" + _statusUpdateInterval + "]";
     }
+
+    /**
+     * Create a {@link JobClientProxy} for submitting requests on behalf of another user.
+     * @param impersonatedUser specifies the authentication principal of the user to be impersonated.
+     * @return a {@link JobClientProxy} that can perform impersonated requests.
+     */
+    public JobClientProxy impersonating(String impersonatedUser) {
```
Comment (pschorf) [inline jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java]: Why can't this return a `JobClientInterface`?

Comment (DaoWen) [reply jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java]: It could, but I wanted to emphasize the fact that this is just a *proxy* over the original JobClient; i.e., all of the mutable internal data structures are shared between the original JobClient object and the impersonation-proxy.

**Comment by dposada** [comment]
Discussed offline, but in summary, we can consider making the impersonating class take in a job client in its constructor and then just delegate to that job client. We could also get rid of the abstract class.

File: jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java [hunk 5]
```diff
@@ -1029,12 +1018,13 @@ public void abort(Collection<UUID> uuids)
...
                 uriBuilder.addParameters(params);
                 httpRequest =  new HttpDelete(uriBuilder.build());
+                addImpersonation(httpRequest, impersonatedUser);
             } catch (URISyntaxException e) {
                 throw releaseAndCreateException(null, null, "Can not submit DELETE request " + params + " via uri " + _jobURI, e);
             }
             HttpResponse httpResponse;
             try {
-                httpResponse = _httpClient.execute(httpRequest);
+                httpResponse = executeWithRetries(httpRequest, 5, 10);
```
Comment (dposada) [inline jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java]: Is this change related to impersonation?

File: jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java [hunk 6]
```diff
@@ -1113,4 +1103,80 @@ public String toString() {
                 + ", _activeUUIDToJob=" + _activeUUIDToJob + ", _batchSubmissionLimit=" + _batchRequestSize
                 + ", _statusUpdateInterval=" + _statusUpdateInterval + "]";
     }
+
+    /**
+     * Create a {@link ImpersonationProxy} for submitting requests on behalf of another user.
+     * @param impersonatedUser specifies the authentication principal of the user to be impersonated.
+     * @return a {@link ImpersonationProxy} that can perform impersonated requests.
+     */
+    public ImpersonationProxy impersonating(String impersonatedUser) {
```
Comment (dposada) [inline jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java]: Can't this return `JobClientInterface`?

File: jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/Constraints.java (hunk 1 above)
Comment (dposada) [reply jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/Constraints.java]: Should we do it in a separate PR?

File: jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/OneToOneConstraint.java (hunk 2 above)
Comment (dposada) [reply jobclient/src/main/java/com/twosigma/cook/jobclient/constraint/OneToOneConstraint.java]: Should we do it in a separate PR?

File: jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java (hunk 5 above)
Comment (DaoWen) [reply jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java]: Yes, because delegating to `executeWithRetries` means I have to handle injecting the impersonation headers in fewer places.

File: jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java (hunk 6 above)
Comment (DaoWen) [reply jobclient/src/main/java/com/twosigma/cook/jobclient/JobClient.java]: Done

**Comment by dposada** [comment]
@pschorf I merged this, but let us know if you have more feedback
